### ISSUE TITLE
Key enrolled, but still getting "Session info error: BadParameter: invalid public key"

### ISSUE THREAD

**Issue by thaya** [iainbullock/tesla-http-proxy-docker#46]
I was able to complete the setup and successfully authenticated Tesla Custom Integrate with Fleet API. I've enrolled my key to my vehicle (clicked approve on the popup and verified that the key showed up under "Locks" in my car). Yet whenever I try to control the car I get "Session info error: BadParameter: invalid public key." Any tips on how to troubleshoot?

Here's are the logs that might be helpful.
`2024-05-14T04:17:08Z [info ] Received POST request for /api/1/vehicles/[REDACTED]/command/set_charge_limit
2024-05-14T04:17:08Z [debug] Executing set_charge_limit on [REDACTED]
2024-05-14T04:17:08Z [info ] Starting dispatcher service...
2024-05-14T04:17:08Z [info ] Requesting session info from DOMAIN_INFOTAINMENT
2024-05-14T04:17:08Z [info ] Requesting session info from DOMAIN_VEHICLE_SECURITY
2024-05-14T04:17:08Z [debug] Sending request to https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/vehicles/[REDACTED]/signed_command: {"routable_message":"MgIIAzoSEhDQiST/GWa9zG0x+IPhyb7vmgMQcpnf/vgPTrzGb1celFjTzXJDCkEEyzD56/22idaXVOVD6qNlMsDEAdQGnQSUG6Xs8aY561mceSXTT51RaKjrqLrdyDWObPy46G58pSOEN3mOefTB/w=="}
2024-05-14T04:17:08Z [debug] Sending request to https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/vehicles/[REDACTED]/signed_command: {"routable_message":"MgIIAjoSEhDUz1gFU0ZcVl12lD4eb75FmgMQJSRzYnLWczg5SHugx7sQbnJDCkEEyzD56/22idaXVOVD6qNlMsDEAdQGnQSUG6Xs8aY561mceSXTT51RaKjrqLrdyDWObPy46G58pSOEN3mOefTB/w=="}
2024-05-14T04:17:09Z [debug] Server returned 200: OK: {"response":"MhISENTPWAVTRlxWXXaUPh5vvkU6AggCegIoAZIDECUkc2Jy1nM4OUh7oMe7EG4="}
2024-05-14T04:17:09Z [warn ] [2524736272d6733839487ba0c7bb106e] Discarding unauthenticated session info
2024-05-14T04:17:09Z [warn ] [2524736272d6733839487ba0c7bb106e] Session info error: BadParameter: invalid public key
2024-05-14T04:17:09Z [error] Returning error Internal Server Error
`

**Comment by thaya** [comment]
I've also tried regenerating keys from scratch and still no luck.

**Comment by iainbullock** [comment]
Please can you share your URL for the public key, as the nginx setup is often the cause of various issues. I don't see a security issue in doing this, as it is public rather than private key. Mine is at https;//tesla.geekleader.co.uk/.well-known/appspecific/com.tesla.3p.public-key.pem

Thanks

**Comment by thaya** [comment]
Thanks for following up! Here's my URL for the public key: https://tesla.chupiebara.com/.well-known/appspecific/com.tesla.3p.public-key.pem

**Comment by iainbullock** [comment]
That's working ok. 

Can you stop the proxy delete everything in /data in the container and also the private key from your car. Try again. Sometimes that works for whatever reason. 

If you still get the same please share your docker-compose.yml and config.sh 

Also all logs from the proxy container since you restarted it

Final thought how old is your car. Some vehicles haven't yet been migrated to the Fleet API

**Comment by thaya** [comment]
Thanks for following up again. I've deleted everything in /data in the container, the private key from the car, the public key, and selfsigned.pem. I tried again and I still get the same error message. My car is a 2023 Model Y. I was able use the Tesla Custom Integration without the Fleet API until about a week ago, so presumably they just switched my car over.

compose.yml
```
version: '3'
volumes:
  tesla_http_proxy:
    external: true
services:
  tesla_http_proxy:
    container_name: tesla-http-proxy
    restart: always
    image: "iainbullock/tesla_http_proxy:latest"

    environment:
      - TZ=America/Los_Angeles
      - CLIENT_ID=4c06207f95a1-4f81-9613-a4b40782bfe0
      - CLIENT_SECRET=[REDACTED]
      - DOMAIN=tesla.chupiebara.com # Public FQDN
      - PROXY_HOST=raspberrypi # Local hostname (not IP) of this docker host
      - REGION=North America, Asia-Pacific

    stdin_open: true
    tty: true

    entrypoint: "/app/run.sh"
    working_dir: /app

    volumes:
      - tesla_http_proxy:/data
      # Webserver root for the $DOMAIN virtual server. Change the path according to your webserver setup. Path must exist or this container won't start
      - type: bind
        source: /config/nginx/htdocs
        target: /share/nginx #Do not modify
      # Path to tesla_http_proxy directory inside /config on Home Assistant instance. Change according to your HA setup. Path must exist or this container won't start
      - type: bind
        source: /config/tesla-http-proxy
        target: /share/home-assistant #Do not modify
      - /data/config.sh:/data/config.sh

    network_mode: bridge
    ports:
      - 4430:443
      - 8099:8099
```

config.sh
```
#!/bin/bash
CLIENT_ID='4c06207f95a1-4f81-9613-a4b40782bfe0'
CLIENT_SECRET='[REDACTED]'
DOMAIN='tesla.chupiebara.com' # Public FQDN
PROXY_HOST='raspberrypi' # Local hostname (not IP) of this docker host
REGION='North America, Asia-Pacific'
OPTIONS_COMPLETE=1
```

See [proxy.log](https://github.com/iainbullock/tesla-http-proxy-docker/files/15343081/proxy.log)

Also confirming that I specified "/config/tesla-http-proxy/selfsigned.pem" into the Tesla Custom Integration auth dialog and it was able to authenticate fine. Reading sensors is fine, but changing any settings throw "invalid public key" as seen in the logs.

**Comment by thaya** [comment]
Not sure if it might matter, but to ssh to the my RPi on the network, I ssh to `raspberrypi.local`, but from within the Home Assistant container, I can only access the hostname `raspberrypi` (without the `.local`). That's why I used `raspberrypi` as the proxyhost in compose.yml and config.sh. I honestly am not sure how to access `raspberrypi.local` from within the HA container.

**Comment by iainbullock** [comment]
Umm not sure. Looking at the logs I see two errors which might provide a clue. Here's the first error:

```
Starting temporary Python app for authentication flow. Delete /data/access_token to force regeneration in the future
[02:30:19] main:INFO: *** Starting Flask server... ***
[02:30:19] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.17.0.2:8099
[02:30:19] werkzeug:INFO: Press CTRL+C to quit
ash: 1: unknown operand
```

The unkown operand error is probably from this line in run.sh:

`python3 /app/run.py --client-id "$CLIENT_ID" --client-secret "$CLIENT_SECRET" --domain "$DOMAIN" --region "$REGION" --proxy-host "$PROXY_HOST"`

Maybe it doesn't like $PROXY_HOST given what you said about the inconsistancy regarding name resolution for your server

The second error in the logs is this:

```
Starting temporary Python app for authentication flow. Delete /data/access_token to force regeneration in the future
[02:33:42] main:INFO: *** Starting Flask server... ***
[02:33:42] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.17.0.2:8099
[02:33:42] werkzeug:INFO: Press CTRL+C to quit
[02:34:49] werkzeug:INFO: 172.17.0.1 - - [17/May/2024 02:34:49] "GET / HTTP/1.0" 200 -
[02:34:49] main:ERROR: 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
```

I think this is coming from this section in run.py:

```
@app.route("/")
def index():
    """Web UI"""
    return render_template(
        "index.html",
        domain=args.domain,
        client_id=args.client_id,
        scopes=SCOPES,
        randomstate=uuid.uuid4().hex,
        randomnonce=uuid.uuid4().hex,
    )
```

Which might mean your Nginx setup isn't right. Please could you share your Nginx config for tesla.chupiebara.com

I tihnk we should get the DNS for your server working consistently as the first step. This was a problem for my setup originally and several others have had problems with DNS.

On my network DNS is done by my router, so I add a static DNS entry mapping my docker server hostname (macmini.home) to its IP address. How you do this it specific to your router (or whatever is providing DNS for your network)

I can then use nslookup to check it resolves correctly both on the host but also inside the containers.

Let me know if you need help with this DNS setup and we can go into more detail

**Comment by ettiennegous** [comment]
@iainbullock how does one delete the private key from your car?

**Comment by iainbullock** [comment]
From the screen in your car select Locks and find the key, press the delete symbol.

@thaya did you dlete the prviate key when you deleted everything above?

**Comment by thaya** [comment]
Thank you for your guidance on the logs @iainbullock! Yes I did delete the private key from the car when I deleted everything above. It looks like my issues was two folds:
1.  "ash: 1: unknown operand" had to do with my docker compose file mounting /data/config.sh as a directory and not a file. I stopped the container, deleted the directory, restarted the container, and viola!
2. The second Invalid Key issue was due to an nginx configuration issue pointing things to the wrong directory as you suspected. The publicly accessible `com.tesla.3p.public-key.pem` was pointing to some other folder with an outdated `com.tesla.3p.public-key.pem` so it was actually giving the correct error message. I fixed that and things are working great!

I also fixed my DNS issue so that everything can reference `raspberrypi.local` per your guidance. I wasn't sure if this would've also created issues but now the Tesla Custom Integration works flawlessly through Fleet API.

Thank you again for all of your support.
